packages/shared/src/hooks/post/useBlockPost.ts

```diff
@@ -0,0 +1,153 @@
+import { useQuery, useQueryClient } from 'react-query';
+import { useMemo } from 'react';
+import { Post } from '../../graphql/posts';
+import { generateStorageKey, StorageTopic } from '../../lib/storage';
+import { useAuthContext } from '../../contexts/AuthContext';
+import { useActions } from '../useActions';
+import { ActionType } from '../../graphql/actions';
+import useTagAndSource from '../useTagAndSource';
+import { Origin } from '../../lib/analytics';
+import {
+  BlockTagSelection,
+  DownvoteBlocked,
+  getBlockedLength,
+} from '../../components/post/block/common';
+
+interface BlockData {
+  showTagsPanel?: boolean;
+  blocked?: DownvoteBlocked;
+}
+
+interface UseBlockPost {
+  data: BlockData;
+  blockedTags: number;
+  onClose(): void;
+  onShowPanel(): void;
+  onDismissPermanently(): void;
+  onReport(): void;
+  onUndo(): void;
+  onBlock(tags: BlockTagSelection, shouldBlockSource: boolean): void;
+}
+
+interface UseBlockPostProps {
+  toastOnSuccess?: boolean;
+}
+
+interface Params {
+  blocks: string[];
+  unblocks: string[];
+}
+
+const getParams = (tags: BlockTagSelection): Params => {
+  const blocks = [];
+  const unblocks = [];
+  Object.entries(tags).forEach(([tag, shouldBlock]) => {
+    const container = shouldBlock ? blocks : unblocks;
+    container.push(tag);
+  });
+
+  return { blocks, unblocks };
+};
+
+export const useBlockPost = (
+  post: Post,
+  { toastOnSuccess }: UseBlockPostProps = {},
+): UseBlockPost => {
+  const { onBlockTags, onUnfollowSource, onUnblockTags, onFollowSource } =
+    useTagAndSource({
+      origin: Origin.TagsFilter,
+      postId: post.id,
+    });
+  const client = useQueryClient();
+  const { user } = useAuthContext();
+  const { checkHasCompleted, completeAction } = useActions();
+  const key = generateStorageKey(
+    StorageTopic.Post,
+    `block:${post?.id}`,
+    user?.id,
+  );
+  const { data } = useQuery<BlockData>(key, () => client.getQueryData(key), {
+    initialData: {},
+  });
+  const setShowTagsPanel = (params: BlockData) =>
+    client.setQueryData<BlockData>(key, (current) => ({
+      ...current,
+      ...params,
+    }));
+
+  const value = useMemo<BlockData>(
+    () => ({
+      ...data,
+      showTagsPanel: checkHasCompleted(ActionType.HideBlockPanel)
+        ? undefined
```

Comment:
`false` would make more sense?

Reply:
`undefined` is more for actually not showing anything. False is for the after-message "5 topics blocked".